<a href="https://colab.research.google.com/github/AsmaBenHmidene/food-sales-predictions/blob/main/Project_1_Part_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# I.Load Data:

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np

In [ ]:
filename = '/content/drive/MyDrive/Coding Dojo/Week2/sales_predictions_2023 (1).csv'
df = pd.read_csv(filename)
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

# II.Cleaning Data:

1.Number of columns and rows

In [ ]:
df.shape

(8523, 12)

2.Data Type:

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


3.Missing value :
item weight and outlet size

In [ ]:
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

Item Weight

For Item wieght, we can fill in missing values using the average.

In [ ]:
df['Item_Weight'].fillna(value =df['Item_Weight'].mean(),inplace=True )
df

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0              FDA15        9.300          Low Fat         0.016047   
1              DRC01        5.920          Regular         0.019278   
2              FDN15       17.500          Low Fat         0.016760   
3              FDX07       19.200          Regular         0.000000   
4              NCD19        8.930          Low Fat         0.000000   
...              ...          ...              ...              ...   
8518           FDF22        6.865          Low Fat         0.056783   
8519           FDS36        8.380          Regular         0.046982   
8520           NCJ29       10.600          Low Fat         0.035186   
8521           FDN46        7.210          Regular         0.145221   
8522           DRG01       14.800          Low Fat         0.044878   

                  Item_Type  Item_MRP Outlet_Identifier  \
0                     Dairy  249.8092            OUT049   
1               Soft Drinks   48.2692            OUT018   
2                      Meat  141.6180            OUT049   
3     Fruits and Vegetables  182.0950            OUT010   
4                 Household   53.8614            OUT013   
...                     ...       ...               ...   
8518            Snack Foods  214.5218            OUT013   
8519           Baking Goods  108.1570            OUT045   
8520     Health and Hygiene   85.1224            OUT035   
8521            Snack Foods  103.1332            OUT018   
8522            Soft Drinks   75.4670            OUT046   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                          1999      Medium               Tier 1   
1                          2009      Medium               Tier 3   
2                          1999      Medium               Tier 1   
3                          1998         NaN               Tier 3   
4                          1987        High               Tier 3   
...                         ...         ...                  ...   
8518                       1987        High               Tier 3   
8519                       2002         NaN               Tier 2   
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  Item_Outlet_Sales  
0     Supermarket Type1          3735.1380  
1     Supermarket Type2           443.4228  
2     Supermarket Type1          2097.2700  
3         Grocery Store           732.3800  
4     Supermarket Type1           994.7052  
...                 ...                ...  
8518  Supermarket Type1          2778.3834  
8519  Supermarket Type1           549.2850  
8520  Supermarket Type1          1193.1136  
8521  Supermarket Type2          1845.5976  
8522  Supermarket Type1           765.6700  

[8523 rows x 12 columns]

Outlet size

missing value represent 28.28% which is more then 2% thus we can't drop rows. We can assume that it's a medium package since that's the most occured category

In [ ]:
df['Outlet_Size'].fillna('Medium', inplace=True)

In [ ]:
df['Outlet_Size'].value_counts(normalize = True).mul(100).round(2)

Medium    61.05
Small     28.02
High      10.94
Name: Outlet_Size, dtype: float64

Conclusion: most sales are done in medium size outlet.

4.Duplicates:
no duplicates

In [ ]:
df.duplicated().sum()

0

5.Irrelevant data for sales of the product in the particular store:
Item_Identifier, item_weight, outlet_identifier, oulet_establishment_year

In [ ]:
df.nunique()

Item_Identifier              1559
Item_Weight                   416
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

In [ ]:
df = df.drop(columns=['Outlet_Establishment_Year', 'Item_Identifier', 'Outlet_Identifier'])
df.head()

Item_Weight Item_Fat_Content  Item_Visibility              Item_Type  \
0         9.30          Low Fat         0.016047                  Dairy   
1         5.92          Regular         0.019278            Soft Drinks   
2        17.50          Low Fat         0.016760                   Meat   
3        19.20          Regular         0.000000  Fruits and Vegetables   
4         8.93          Low Fat         0.000000              Household   

   Item_MRP Outlet_Size Outlet_Location_Type        Outlet_Type  \
0  249.8092      Medium               Tier 1  Supermarket Type1   
1   48.2692      Medium               Tier 3  Supermarket Type2   
2  141.6180      Medium               Tier 1  Supermarket Type1   
3  182.0950      Medium               Tier 3      Grocery Store   
4   53.8614        High               Tier 3  Supermarket Type1   

   Item_Outlet_Sales  
0          3735.1380  
1           443.4228  
2          2097.2700  
3           732.3800  
4           994.7052

check if Item_Visibility is relevent data

In [ ]:
filter_visibility= df['Item_Visibility'] != 0.00
df_filter = df.loc[filter_visibility, :]
df_filter.head()

Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  Item_MRP  \
0     9.300000          Low Fat         0.016047        Dairy  249.8092   
1     5.920000          Regular         0.019278  Soft Drinks   48.2692   
2    17.500000          Low Fat         0.016760         Meat  141.6180   
6    13.650000          Regular         0.012741  Snack Foods   57.6588   
7    12.857645          Low Fat         0.127470  Snack Foods  107.7622   

  Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0      Medium               Tier 1  Supermarket Type1          3735.1380  
1      Medium               Tier 3  Supermarket Type2           443.4228  
2      Medium               Tier 1  Supermarket Type1          2097.2700  
6        High               Tier 3  Supermarket Type1           343.5528  
7      Medium               Tier 3  Supermarket Type3          4022.7636

In [ ]:
df_filter['Item_Visibility'].describe().round(2)

count    7997.00
mean        0.07
std         0.05
min         0.00
25%         0.03
50%         0.06
75%         0.10
max         0.33
Name: Item_Visibility, dtype: float64

In [ ]:
df_filter['Item_Visibility'].idxmax()

854

In [ ]:
df_filter.iloc[854]['Item_Type']

'Fruits and Vegetables'

In [ ]:
df_filter['Item_Outlet_Sales'].idxmax()

7188

In [ ]:
df_filter.iloc[7188]['Item_Type']

'Health and Hygiene'

In [ ]:
df_filter['Item_Visibility'].idxmin()

7551

In [ ]:
df_filter.iloc[7551]['Item_Type']

'Fruits and Vegetables'

In [ ]:
df_filter.groupby('Item_Type')['Item_Visibility'].mean().round(2)

Item_Type
Baking Goods             0.07
Breads                   0.07
Breakfast                0.09
Canned                   0.07
Dairy                    0.08
Frozen Foods             0.07
Fruits and Vegetables    0.07
Hard Drinks              0.07
Health and Hygiene       0.06
Household                0.06
Meat                     0.07
Others                   0.06
Seafood                  0.08
Snack Foods              0.07
Soft Drinks              0.07
Starchy Foods            0.08
Name: Item_Visibility, dtype: float64

In [ ]:
df_filter.groupby('Item_Type')['Item_Visibility'].mean().round(2).value_counts()

0.07    9
0.08    3
0.06    3
0.09    1
Name: Item_Visibility, dtype: int64

In [ ]:
df_filter.groupby('Outlet_Type')['Item_Visibility'].mean().round(2)

Outlet_Type
Grocery Store        0.11
Supermarket Type1    0.06
Supermarket Type2    0.07
Supermarket Type3    0.06
Name: Item_Visibility, dtype: float64

Conclusion: Item_visibility is very variable according to different parameters: the most visible item is the least visible as well and is not the most selled one. we will drop it as well

In [ ]:
df = df.drop(columns=['Item_Visibility'])
df.head()

Item_Weight Item_Fat_Content              Item_Type  Item_MRP Outlet_Size  \
0         9.30          Low Fat                  Dairy  249.8092      Medium   
1         5.92          Regular            Soft Drinks   48.2692      Medium   
2        17.50          Low Fat                   Meat  141.6180      Medium   
3        19.20          Regular  Fruits and Vegetables  182.0950      Medium   
4         8.93          Low Fat              Household   53.8614        High   

  Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0               Tier 1  Supermarket Type1          3735.1380  
1               Tier 3  Supermarket Type2           443.4228  
2               Tier 1  Supermarket Type1          2097.2700  
3               Tier 3      Grocery Store           732.3800  
4               Tier 3  Supermarket Type1           994.7052

6.inconsistent categories of data

In [ ]:
df['Item_Fat_Content'].value_counts(normalize = True).mul(100).round(2)

Low Fat    59.71
Regular    33.90
LF          3.71
reg         1.37
low fat     1.31
Name: Item_Fat_Content, dtype: float64

In [ ]:
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('LF','Low Fat').replace('low fat','Low Fat').replace('reg','Regular')
df['Item_Fat_Content'].value_counts(normalize = True).mul(100).round(2)

Low Fat    64.73
Regular    35.27
Name: Item_Fat_Content, dtype: float64

In [ ]:
df.groupby('Item_Fat_Content')['Item_Outlet_Sales'].mean().round(2)

Item_Fat_Content
Low Fat    2157.71
Regular    2224.56
Name: Item_Outlet_Sales, dtype: float64

In [ ]:
filter_FatContent= df['Item_Fat_Content'] == "Regular"
df_filter_FatContent = df.loc[filter_FatContent, :]
df_filter_FatContent.head()

Item_Weight Item_Fat_Content              Item_Type  Item_MRP Outlet_Size  \
1        5.920          Regular            Soft Drinks   48.2692      Medium   
3       19.200          Regular  Fruits and Vegetables  182.0950      Medium   
5       10.395          Regular           Baking Goods   51.4008      Medium   
6       13.650          Regular            Snack Foods   57.6588        High   
8       16.200          Regular           Frozen Foods   96.9726      Medium   

  Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
1               Tier 3  Supermarket Type2           443.4228  
3               Tier 3      Grocery Store           732.3800  
5               Tier 3  Supermarket Type2           556.6088  
6               Tier 3  Supermarket Type1           343.5528  
8               Tier 2  Supermarket Type1          1076.5986

In [ ]:
df_filter_FatContent.groupby('Item_Type')['Item_Outlet_Sales'].mean().round(2).idxmax()

'Seafood'

In [ ]:
df_filter_FatContent.groupby('Item_Type')['Item_Outlet_Sales'].mean().round(2).idxmin()

'Soft Drinks'

In [ ]:
filter_FatContent2= df['Item_Fat_Content'] == "Low Fat"
df_filter_FatContent2 = df.loc[filter_FatContent2, :]
df_filter_FatContent2.head()

Item_Weight Item_Fat_Content              Item_Type  Item_MRP Outlet_Size  \
0      9.300000          Low Fat                  Dairy  249.8092      Medium   
2     17.500000          Low Fat                   Meat  141.6180      Medium   
4      8.930000          Low Fat              Household   53.8614        High   
7     12.857645          Low Fat            Snack Foods  107.7622      Medium   
10    11.800000          Low Fat  Fruits and Vegetables   45.5402      Medium   

   Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0                Tier 1  Supermarket Type1          3735.1380  
2                Tier 1  Supermarket Type1          2097.2700  
4                Tier 3  Supermarket Type1           994.7052  
7                Tier 3  Supermarket Type3          4022.7636  
10               Tier 1  Supermarket Type1          1516.0266

In [ ]:
df_filter_FatContent2.groupby('Item_Type')['Item_Outlet_Sales'].mean().round(2).idxmax()

'Starchy Foods'

In [ ]:
df_filter_FatContent2.groupby('Item_Type')['Item_Outlet_Sales'].mean().round(2).idxmin()

'Baking Goods'

conclusion: regular products represent the principle sales and more specifically seafood.
For Low fat item, starchy foods represent the most selled product.

#Analyse and final conclusion:

Effect of outlet type on sales

In [ ]:
df['Outlet_Type'].value_counts(normalize = True).mul(100).round(2)

Supermarket Type1    65.43
Grocery Store        12.71
Supermarket Type3    10.97
Supermarket Type2    10.89
Name: Outlet_Type, dtype: float64

In [ ]:
df.groupby('Outlet_Type')['Item_Outlet_Sales'].mean().round(2)

Outlet_Type
Grocery Store         339.83
Supermarket Type1    2316.18
Supermarket Type2    1995.50
Supermarket Type3    3694.04
Name: Item_Outlet_Sales, dtype: float64

Conclusion: Supermarket Type3 represent only 10.97% however they show the highest sales average.

Effect of the oulet location type on sales

In [ ]:
df['Outlet_Location_Type'].value_counts(normalize = True).mul(100).round(2)

Tier 3    39.31
Tier 2    32.68
Tier 1    28.02
Name: Outlet_Location_Type, dtype: float64

In [ ]:
df.groupby('Outlet_Location_Type')['Item_Outlet_Sales'].mean().round(2)

Outlet_Location_Type
Tier 1    1876.91
Tier 2    2323.99
Tier 3    2279.63
Name: Item_Outlet_Sales, dtype: float64

Conclusion: Outlets in location Tier 2 and Tier 3 show the highest sales average.

#Final conclusion:

The properties of products and outlets that play crucial roles in increasing sales:
- Outlet_location_type: Tier 2 and Tier 3
- Outlet_Type: Supermarket Type3
- Outlet_Size: medium size and high size outlets
- Item_Fat_Content: regular products represent the principle sales and more specifically seafood. For Low fat item, starchy foods represent the most selled product.